### PR TITLE
fix(circularDeps): correct $http circular dependency

### PR DESCRIPTION
**Pull request by CITguy**
JIRA: https://jira.rax.io/browse/FRMW-1312

### LGTMs
- [x] Dev LGTM
- [x] Dev+Vidyo LGTM
- [x] QE LGTM

### CODE REVIEW

**Comment by CITguy** [comment]
Part of this PR was to get proper deprecations in place for all of the deprecated services involved. As such, we split rxAuth into two services. 

* `rxIdentity` for Identity/`$resource` logic
* `rxSession` for session management and utility functions

3.3.x | 3.4.x | 3.4.3
----- | ----- | -----
Auth | rxAuth | rxIdentity, rxSession
Identity | rxAuth | rxIdentity
Session | rxAuth | rxSession

File: src/utilities/Session/docs/examples/Session.simple.html [hunk 1]
```diff
@@ -1,3 +0,0 @@
-<div ng-controller="SessionSimpleCtrl">
-    <button ng-click="isAuthenticated()" class="button">Are You Authenticated?</button>
-</div>
```
Comment (CITguy) [inline src/utilities/Session/docs/examples/Session.simple.html]: Example moved to `src/utilities/rxSession/docs/examples`

File: src/utilities/rxAuth/docs/examples/rxAuth.simple.js [hunk 2]
```diff
@@ -1,10 +0,0 @@
-angular.module('demoApp')
-.controller('rxAuthSimpleCtrl', function ($scope, $window, rxAuth) {
-    $scope.hasRole = function () {
-        $window.alert('Has "superhero" Role? : ' + rxAuth.hasRole('superhero'));
-    };
-
-    $scope.isAuthenticated = function () {
-        $window.alert('Is Authenticated? : ' + rxAuth.isAuthenticated());
-    };
-});
```
Comment (CITguy) [inline src/utilities/rxAuth/docs/examples/rxAuth.simple.js]: Example combined wit `src/utilities/rxSession/docs/examples`

File: src/utilities/rxAuth/scripts/rxAuth.js [hunk 3]
```diff
@@ -1,263 +1,25 @@
...
-     * @methodOf utilities.service:rxAuth
-     * @description Checks if user has _every_ role in given list.
-     * @param {String[]} roles List of roles to check against
-     * @returns {Boolean} True if user has _every_ role in given list; otherwise, false.
-     */
-    svc.hasAllRoles = function (roles) {
-        return checkRoles(roles, _.every);
-    };
+    _.assign(svc, rxIdentity);
+    _.assign(svc, rxSession);
```
Comment (CITguy) [inline src/utilities/rxAuth/scripts/rxAuth.js]: Logic is functionally identical to `Auth` service.

File: src/utilities/rxAuth/scripts/rxAuth.spec.js [hunk 4]
```diff
@@ -19,159 +19,69 @@ describe('utilities:rxAuth', function () {
...
 
         it('hasRole() should validate user has role via rxAuth.hasRole', function () {
-            expect(auth.hasRole('admin')).to.be.true;
-            expect(auth.hasRole('fakeRole')).to.be.false;
-            expect(auth.getRoles).to.be.called;
-        });
-    });
-});
-
-describe('utilities:Auth (DEPRECATED)', function () {
```
Comment (CITguy) [inline src/utilities/rxAuth/scripts/rxAuth.spec.js]: `Auth` specs moved to `src/utilities/Auth/scripts/Auth.spec.js` where they belong.

File: src/utilities/rxAuth/scripts/rxAuth.js [hunk 5]
```diff
@@ -1,263 +1,25 @@
...
-})
-
-/**
- * @deprecated
- * Please use rxAuth instead. This item will be removed on the 4.0.0 release.
- * @ngdoc service
- * @name utilities.service:Auth
- * @requires utilities.service:rxAuth
- */
-.service('Auth', function (rxAuth) {
```
Comment (CITguy) [inline src/utilities/rxAuth/scripts/rxAuth.js]: `Auth` service moved to `src/utilities/Auth/scripts/Auth.js` where it belongs.

File: src/utilities/rxTokenInterceptor/scripts/rxTokenInterceptor.js [hunk 6]
```diff
@@ -36,46 +36,7 @@ angular.module('encore.ui.utilities')
...
-    };
-})
-
-/**
- * @deprecated
- * Please use rxTokenInterceptor instead. This item will be removed on the 4.0.0 release.
- * @ngdoc service
- * @name utilities.service:TokenInterceptor
- */
-.provider('TokenInterceptor', function () {
```
Comment (CITguy) [inline src/utilities/rxTokenInterceptor/scripts/rxTokenInterceptor.js]: `TokenInterceptor` moved to `src/utilities/TokenInterceptor` where it belongs.

File: src/utilities/rxTokenInterceptor/scripts/rxTokenInterceptor.spec.js [hunk 7]
```diff
@@ -16,70 +16,8 @@ describe('provider:rxTokenInterceptor', function () {
...
-        var config = {
-            headers: {},
-            url: 'https://rackspace.com/rackcdn.com'
-        };
-        interceptor.request(config);
-        expect(config.headers).to.not.be.empty;
-    });
-});
-
-describe('provider:TokenInterceptor (DEPRECATED)', function () {
```
Comment (CITguy) [inline src/utilities/rxTokenInterceptor/scripts/rxTokenInterceptor.spec.js]: `TokenInterceptor` specs moved to `src/utilities/TokenInterceptor` where they belong.

File: src/utilities/rxUnauthorizedInterceptor/scripts/rxUnauthorizedInterceptor.js [hunk 8]
```diff
@@ -43,19 +43,4 @@ angular.module('encore.ui.utilities')
...
-})
-
-/**
- * @deprecated
- * Please use rxUnauthorizedInterceptor instead. This item will be removed on the 4.0.0 release.
- * @ngdoc service
- * @name utilities.service:UnauthorizedInterceptor
- * @requires utilities.service:rxUnauthorizedInterceptor
- */
-.service('UnauthorizedInterceptor', function (rxUnauthorizedInterceptor) {
```
Comment (CITguy) [inline src/utilities/rxUnauthorizedInterceptor/scripts/rxUnauthorizedInterceptor.js]: `UnauthorizedInterceptor` moved to `src/utilities/UnauthorizedInterceptor` where it belongs.

File: src/utilities/rxUnauthorizedInterceptor/scripts/rxUnauthorizedInterceptor.spec.js [hunk 9]
```diff
@@ -38,86 +38,7 @@ describe('service:rxUnauthorizedInterceptor', function () {
...
-
-    it('Interceptor should generate redirect with base path specified', function () {
-        sinon.stub(interceptor, 'redirectPath').returns(cases.fullPath);
-
-        interceptor.redirect('/mario');
-        expect(mockWindow.location).to.contain('/mario');
-    });
-});
-
-describe('service:UnauthorizedInterceptor (DEPRECATED)', function () {
```
Comment (CITguy) [inline src/utilities/rxUnauthorizedInterceptor/scripts/rxUnauthorizedInterceptor.spec.js]: `UnauthorizedInterceptor` specs moved to `src/utilities/UnauthorizedInterceptor` where they belong.

File: src/utilities/TokenInterceptor/scripts/TokenInterceptor.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+ */
+.provider('TokenInterceptor', function () {
+    var exclusionList = this.exclusionList = [ 'rackcdn.com' ];
+
+    this.$get = function (rxSession, $document) {
+        console.warn (
+            'DEPRECATED: TokenInterceptor - Please use rxTokenInterceptor. ' +
+            'TokenInterceptor will be removed in EncoreUI 4.0.0'
+        );
+        var url = $document[0].createElement('a');
```
Comment (flazz) [inline src/utilities/TokenInterceptor/scripts/TokenInterceptor.js]: is `$document[0]` always guaranteed to exist?

Comment (CITguy) [inline src/utilities/TokenInterceptor/scripts/TokenInterceptor.js]: Yes

File: src/utilities/rxSession/scripts/rxSession.js [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+         * @methodOf utilities.service:rxSession
+         * @description Dot walks the token without throwing an error.
+         * If key exists, returns value otherwise returns undefined.
+         * @param {Function} key callback
+         * @returns {String} Key value
+         */
+        svc.getByKey = function (key) {
+            var tokenValue,
+                token = svc.getToken(),
+                keys = key ? key.split('.') : undefined;
```
Comment (flazz) [inline src/utilities/rxSession/scripts/rxSession.js]: why would key be falsy? what does splitting a function do?

**Comment by cathysiller** [comment]
Dev LGTM

File: src/utilities/rxSession/scripts/rxSession.js (hunk 11 above)
Comment (CITguy) [inline src/utilities/rxSession/scripts/rxSession.js]: An empty string is considered falsey in JavaScript. The splitting allows for "dot-walking" the object. This is pre-existing logic that already existed in `Session` (then `rxAuth`) before being moved to `rxSession`.

File: src/utilities/rxSession/scripts/rxSession.js [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+    function rxSessionFactory (rxLocalStorage) {
+        var TOKEN_ID = 'encoreSessionToken';
+        var svc = {};
+
+        /**
+         * @ngdoc function
+         * @name rxSession.getByKey
+         * @methodOf utilities.service:rxSession
+         * @description Dot walks the token without throwing an error.
+         * If key exists, returns value otherwise returns undefined.
```
Comment (Droogans) [inline src/utilities/rxSession/scripts/rxSession.js]: Doesn't [`lodash.get`](https://lodash.com/docs/4.17.4#get) do about this same thing now?

May be able to simplify a lot of this code.

File: src/utilities/rxSession/scripts/rxSession.spec.js [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+        });
+    });//storeToken()
+
+    describe('logout()', function () {
+        beforeEach(function () {
+            rxSession.logout('encoreToken');
+        });
+
+        it('should call rxLocalStorage.removeItem()', function () {
+            expect(rxLocalStorage.removeItem).to.be.called;
```
Comment (Droogans) [inline src/utilities/rxSession/scripts/rxSession.spec.js]: Are we actually doing this? I know the method was called, but there isn't a way to actually "logout" anymore.

Should this be removed?

Comment (CITguy) [inline src/utilities/rxSession/scripts/rxSession.spec.js]: We can do that for the 4.0 release, but this PR is to correct the circular dependency in 3.4.x

File: src/utilities/rxSession/scripts/rxSession.js (hunk 12 above)
Comment (CITguy) [inline src/utilities/rxSession/scripts/rxSession.js]: Oooh! Definitely have to update to using that in 4.x! I was not aware that lodash had that built in.

File: src/utilities/rxSession/scripts/rxSession.spec.js (hunk 13 above)
Comment (CITguy) [inline src/utilities/rxSession/scripts/rxSession.spec.js]: Also, the `logout()` function is still called internally by `rxUnauthorizedInterceptor`.

**Pull request merged**